In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt



**Based on california housing dataset answer the Questions that follow**

##Q1
What is the shape of feature matrix for the dataset?

How many missing values are present in the dataset?


In [ ]:
from sklearn.datasets import fetch_california_housing
# (data,target) = fetch_california_housing(return_X_y= True)

dataset = fetch_california_housing(as_frame = True)
print("shape of the dataset" ,dataset.data.shape)
print("No null values","\n" ,dataset.data.isna().count())



shape of the dataset (20640, 8)
No null values 
 MedInc        20640
HouseAge      20640
AveRooms      20640
AveBedrms     20640
Population    20640
AveOccup      20640
Latitude      20640
Longitude     20640
dtype: int64


##Q2

Instructions: split the california housing dataset into train and test sets.

[Hint: use appropriate API with random_state=0, shuffle=False and test_size=0.2]

- What is the value of house age in third sample of test set?

- What is the value of Population in the first sample of training set?

- What is the mean of output label in training set?

- What is the median of output label in test set?


In [55]:
from sklearn.model_selection import train_test_split

X,y = fetch_california_housing(return_X_y = True)
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=0,shuffle=False)
# using array indexing

#1.
print("value of house age in third sample of test set", X_test[2][1])

#2.
print("value of Population in the first sample of training set", X_train[0][4])

#3.
print(" mean of output label in training set", y_train.mean())

#4.
print(" median of output label in test set", np.median( y_test))



value of house age in third sample of test set 28.0
value of Population in the first sample of training set 322.0
 mean of output label in training set 2.02067031310562
 median of output label in test set 2.1475


Let's explore some of the popular attributes and methods of **`sklearn.linear_model.LinearRegression`** in simple terms:

1. **`coef_` (Estimated Coefficients)**:
   - Imagine you're building a house price prediction model. The **`coef_`** attribute tells you how much each feature (like the number of bedrooms or square footage) contributes to the predicted house price.
   - For example, if the coefficient for the number of bedrooms is 50, it means that for every additional bedroom, the predicted house price goes up by $50 (or whatever currency you're using).

2. **`intercept_` (Independent Term)**:
   - Think of this as the **base price** of the house. Even if all the features (like bedrooms, square footage, etc.) are zero, there's still a minimum price. That's the **`intercept_`**.
   - If the intercept is 10,000, it means the house price starts at $10,000 even before considering any features.

3. **`score(X, y)` (Model Accuracy)**:
   - This method tells you how well your model is doing. It's like a report card for your predictions.
   - If the score is 1.0, it means your model predicts perfectly (which rarely happens). A lower score means there's room for improvement.

4. **`predict(X_new)` (Making Predictions)**:
   - Suppose you have a new house with specific features (like bedrooms and square footage). You can use this method to predict its price.
   - For example, if you input the number of bedrooms and square footage, it will give you an estimated house price.

5. **`fit(X, y)` (Training the Model)**:
   - This is where the magic happens! You give it your training data (features **X** and actual prices **y**), and it learns how to make predictions.
   - It figures out the best coefficients and intercept to minimize prediction errors.

6. **`rank` and `singular` (Matrix Properties)**:
   - These attributes are like behind-the-scenes details. They tell you about the mathematical properties of the data matrix.
   - **`rank`** tells you how many important features your model can learn from.
   - **`singular`** values are related to the stability of the model.



##Q3
If we use StandardScaler for preprocessing and LinearRegression for fitting the model with training set obtained in previous questions, what are the coefficients obtained in the predicted model?

    Note: Use methods and objects with default parameters.

[ 0.7296 , 0.2187, -0.3655, 0.4056, -0.0045 , -0.0493, -0.9998, -0.8705 ]

[ 0.9296 , 0.2287, -0.3655, 0.4056, -0.0045 , -0.0393, -0.8998, -0.8705 ]

[ 0.2961, 0.1875, -0.6552, 0.30569623, -0.0045 , -0.0393, -0.8998, -0.8705 ]

[ 0.8381, 0.1203, -0.3213, 0.3645, -0.0017,-0.0452, -0.8799, -0.8263]

In [57]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression

scaler = StandardScaler()
X_train_norm =scaler.fit_transform(X_train)
X_test_norm = scaler.transform(X_test)

model = LinearRegression()
model.fit(X_train_norm,y_train)

print(model.score(X_test_norm, y_test))
print(model.coef_)

0.6605140591531986
[ 0.83817967  0.12032096 -0.32135834  0.36456599 -0.00170674 -0.04524631
 -0.87994217 -0.82634126]


##Q4

If we use StandardScaler for preprocessing and LinearRegression for fitting the model with training set used in previous question, what is the bias term predicted by the model?

    Note: Use methods and objects with default parameters

1.06855

2.02067

3.06855

-2.06855

In [58]:
model.intercept_

2.020670313104707

``**bias**``

Certainly! Let's dive into the concept of the **bias term** in a way that's easy to understand.

Imagine you're trying to predict something, like how much a house will cost based on its size, number of bedrooms, and other features. You decide to use a mathematical model to make this prediction. One common type of model is called **linear regression**.

Now, let's break it down:

1. **Linear Regression**:
   - Linear regression is like drawing a straight line through a scatterplot of data points. It's a simple way to find a relationship between an input (like house size) and an output (like house price).
   - The goal is to find the best-fitting line that predicts the output based on the input features.

2. **The Line Equation**:
   - In linear regression, we use an equation for the line. It looks like this:
     ```
     predicted_output = slope * input_feature + bias
     ```
   - Here:
     - **Slope**: Represents how much the output changes for a one-unit change in the input feature. It's like the steepness of the line.
     - **Bias**: This is what we're talking about! The bias is like the starting point of the line. It's where the line crosses the vertical axis (y-axis).
     - Together, the slope and bias determine the position and angle of the line.

3. **Why Do We Need a Bias?**:
   - Imagine you're predicting house prices. If you didn't have a bias, the line would always pass through the origin (0,0) on the graph.
   - But in reality, houses have a minimum price even if they're tiny. That's where the bias comes in!
   - The bias allows the line to shift up or down, adjusting for the minimum price (or any other starting point) that makes sense.

4. **Analogies**:
   - Think of the bias as the **base price** for houses. Even if a house has zero bedrooms (input feature), it still costs something because of the land, location, and other factors.
   - Or imagine you're baking cookies. The bias is like the **fixed cost** of ingredients (flour, sugar, etc.) before you even start adding chocolate chips or nuts (the slope).

5. **In Summary**:
   - The bias helps our model account for things that don't depend on the input features. It's like the starting point or the minimum value.
   - Without the bias, our predictions might be way off because we'd be ignoring this essential starting point.

So, next time you hear about the bias term in machine learning, remember it's like the base price or the fixed cost that ensures our model starts from the right place. 🏠📏


##Q5
If we use StandardScaler for preprocessing and LinearRegression for fitting the model with training set obtained from following code snippet:


Assume that (X,y) is the california housing dataset.

- What is the R squared value for predictions obtained using test set features?
- what is the root mean squared error value for predictions obtained using test set features?


In [87]:
from sklearn.metrics import r2_score, mean_squared_error
y_pred= model.predict(X_test_norm) #prediction on normalized data
r2 = r2_score(y_test, y_pred)
print("R squared value for predictions obtained using test set features",r2)


MSE= mean_squared_error(y_test, y_pred)
print("root mean squared error value for predictions obtained using test set features", np.sqrt(MSE))


R squared value for predictions obtained using test set features 0.6605140591531986
root mean squared error value for predictions obtained using test set features 0.7033383507521884


##Q6

- Explained Variance score
- Max Error
- Mean Absolute Error
- Mean Squared Error


In [90]:
from sklearn.metrics import explained_variance_score
EVS= explained_variance_score(y_test, y_pred)
print("EVS", EVS)

from sklearn.metrics import max_error
ME= max_error(y_test,y_pred)
print("ME", ME)

from sklearn.metrics import mean_absolute_error
MAE= mean_absolute_error(y_test, y_pred)
print("MAE", MAE)


print("MSE", MSE)

EVS 0.6605500501742696
ME 7.2604532929585535
MAE 0.5168526993787044
MSE 0.4946848356388085


##Q7

Perform SGD Regression on the given (scaled using StandardScalar()) dataset, using default hyperparameters and random_state=0.

- What is the bias term?
- What are the coefficients in predicted model?

In [95]:
from sklearn.linear_model import SGDRegressor
sgd = SGDRegressor(random_state=0)
sgd.fit(X_train_norm,y_train)
print(sgd.coef_,"\n")
print(sgd.intercept_)

[ 0.84046697  0.112331   -0.41213039  0.21595971 -0.01781887 -0.01480892
 -0.87394103 -0.83913104] 

[2.01123921]
